src/browser/modules/D3Visualization/components/Graph.tsx

```diff
@@ -52,141 +49,81 @@ type GraphProps = {
-      // so, the same graph will be calculated twice
-      // for now, we added conditions to control which parts to be drawn to avoid repeated rendering
-      // the function need to be refactored in a better way in the future
-      this.initGraphView()
-      this.graph && this.props.setGraph && this.props.setGraph(this.graph)
-      this.props.getAutoCompleteCallback &&
-        this.props.getAutoCompleteCallback(this.addInternalRelationships)
-      this.props.assignVisElement &&
-        this.props.assignVisElement(this.svgElement, this.graphView)
-    }
-  }
-
-  initGraphView(): void {
-    if (!this.graphView) {
-      const NeoConstructor = GraphView
-      const measureSize = () => ({
-        width: this.svgElement?.parentElement.clientWidth,
-        height: this.svgElement?.parentElement.clientHeight
-      })
-      this.graph = createGraph(this.props.nodes, this.props.relationships)
-      this.graphView = new NeoConstructor(
-        this.svgElement,
-        measureSize,
-        this.graph,
-        this.props.graphStyle
-      )
-      this.graphEventHandler = new GraphEventHandler(
-        this.graph,
-        this.graphView,
-        this.props.getNodeNeighbours,
-        this.props.onItemMouseOver,
-        this.props.onItemSelect,
-        this.props.onGraphModelChange
+    props.setGraph && props.setGraph(graph)
+    props.getAutoCompleteCallback &&
+      props.getAutoCompleteCallback(
+        (internalRelationships: BasicRelationship[]) => {
+          graph.addInternalRelationships(
+            mapRelationships(internalRelationships, graph)
+          )
+          props.onGraphModelChange(getGraphStats(graph))
+          graphView?.update({ updateNodes: false, updateRelationships: true })
+          graphEventHandler?.onItemMouseOut()
+        }
       )
-      this.graphEventHandler.bindEventHandlers()
-      this.props.onGraphModelChange(getGraphStats(this.graph))
-      this.graphView.resize()
-      this.graphView.update({
-        updateNodes: true,
-        updateRelationships: false,
-        precompute: true
-      })
-    }
-  }
+    props.assignVisElement && props.assignVisElement(svgElement, graphView)
 
-  addInternalRelationships = (
-    internalRelationships: BasicRelationship[]
-  ): void => {
-    if (this.graph) {
-      this.graph.addInternalRelationships(
-        mapRelationships(internalRelationships, this.graph)
-      )
-      this.props.onGraphModelChange(getGraphStats(this.graph))
-      this.graphView?.update({ updateNodes: false, updateRelationships: true })
-      this.graphEventHandler?.onItemMouseOut()
+    setGraphView(graphView)
+  }, [svgElement.current])
+
+  useEffect(() => graphView?.init(), [svgElement.current, props.styleVersion])
+  useEffect(() => graphView?.resize(), [svgElement.current, props.isFullscreen])
+
+  const zoomInClicked = (): void => {
+    if (graphView) {
+      graphView.zoomIn()
```

Comment:
When I dig into the details of zoom while working on the zoom interactions card I actually found a way to get the current zoom level. However you could not get it at anytime, but when a zoom event just has happened. So it can be extracted in the function which is called on all zoom events (``.on('zoom', (e: D3ZoomEvent<SVGGElement, unknown>)...``), and get the current scale with e.transform.k. So I guess we could save that value to the state and from there we can access it whenever we want.

Reply:
Nice! That's good to know. I think it would be best to keep it for a separate PR, given that there are already a lot of changes here.